### ISSUE TITLE
Containerize all the things

### ISSUE THREAD

**Issue by jdegoes** [quasar-analytics/quasar#1730]
The goals of this ticket are as follows:

1. **Deprecate Travis**. Rip out all the infrastructure that depends on Travis. When the ticket is complete, `.travis.yml`, if it is still utilized at all, should be a skeleton of its former self, merely calling shell scripts to build and test. This will pave the way for migrating from Travis to another CI that provides much faster, more stable builds.
2. **Improve Determinism**. Remove the maximum number of hidden dependencies, points of failure, and points of indeterminism. Developers should be able to check out the code base and run all the tests without having to install or configure external dependencies. Moreover, such tests should succeed all the time if they succeed once, and fail all the time if they fail once, up to the determinism limitations of the tests themselves.
3. **Provide Flexibility**. Allow running just the tests for a given backend (plus core tests). This can be utilized by developers working on a specific backend, but also by testing infrastructure that's capable of running the same build in parallel over a matrix of configuration options (i.e. Travis' matrix, but without Travis).

Currently, the best route to obtaining these goals seems to involve containerization. Specifically, if for every connector, we have a mechanism to build a container that has all dependencies and can run all tests for the connector, then we can completely isolate different connectors, parallelize on arbitrary (non-Travis) building & testing infrastructure, and maximize determinism.

To achieve these goals, the following steps are needed (and may be revised as necessary):

- [x] Create Dockerfiles to encode the recipes necessary for building dependencies together with test infrastructure.
- [ ] Create shell scripts or sbt tasks to run tests for a specified connector.
- [x] Simplifying Travis to utilize the new machinery.
- [ ] Downloading or at least detecting the presence of Docker / rkt to enable bootstrapping.
- [ ] ???

**Comment by zeeiyerWork** [comment]
@nicflores , @jdegoes -> Fyi , for now the goal is not necessarily to rip out Travis. Agreed that we should minimize any dependencies, so we can plug/play any CI with minimal fuss. We are trying to fix things without necessarily switching the CI, via working with Travis support on possible options and upgrades. More can be gleaned [here](https://docs.google.com/document/d/1DwabbFo-vPfKjMnHT1VmWO-3dkjQEAnaADFIda1z83k/.). However, we do want to Dockerize and this would be a great way for you to add value immediately.

**Comment by zeeiyerWork** [comment]
Fyi -> https://docs.travis-ci.com/user/docker/

**Comment by nicflores** [comment]
Here are some thoughts I'd like to get out there regarding this task.

Several ideas come to mind.

**First**: In order to have developers run isolated tests against containerized connectors they will need to have docker installed locally on their development machines. Depending on the OS the developer is running they would need to go through an installation process outlined [here](https://docs.docker.com/engine/installation/).

This defeats the idea that all the developer should have to do run commands like `sbt testMy<ConnectorName>` and not have to worry about mucking with their dev box.

**Second**: There's another tool called [Docker Compose](https://docs.docker.com/compose/) that allows one to define an entire dockerized environment in `yml`. Here's a good [example](https://docs.docker.com/compose/wordpress/). Again, though, this would require developers to have Docker Compose installed on their dev box.

**Third**: I happened to come upon an [example](https://docs.travis-ci.com/user/docker/#Using-Docker-Compose) in the Travis-CI docs of how to use Docker Compose. I'll be experimenting with this on one of my own repos. This functionality could bring down build times and also possibly avoid having  developers install docker on their machines if we are willing to interface with travis-ci via the command line tool. For example, we could trigger integration tests on branches containing a docker-compose.yml file using something similar to this [example](https://docs.travis-ci.com/user/triggering-builds/).

If you ya'll have any feedback on this I'd appreciate it. In the mean time I'll keep moving forward with the tasks above beginning with creating dockerfiles for the various databases. I'll be starting with the various versions of MongoDB.

**Comment by nicflores** [comment]
I gave @drostron a quick demo of docker compose yesterday and demonstrating that using a command like `docker-compose up mongo26 mongo30` bring ups those specific databases in containers locally. A plain `docker-compose up` would being up every service defined in the `docker-compose.yml` file. Dave and I agreed that if developers had docker installed this would help ease db setup for integration tests.

The demo pulled docker images from docker hub so the next step is to make docker compose configure each database as we needed. This my next task.

**Comment by jdegoes** [comment]
Ultimately it'd be nice to have an SBT task to pull down Docker / Docker Compose.

**Comment by nicflores** [comment]
@jdegoes, @drostron and I have been bouncing ideas off each other on that topic. I found this [sbt](https://github.com/Tapad/sbt-docker-compose) plugin which might help once a developer has docker and docker-compose installed. As for getting docker on developer dev machines @drostron found this linux [script](https://get.docker.com) which may help. I'm still on the hunt for something similar for mac and windows.

On another note I managed to figure out how to get around this docker couchbase issue I was having:

```
{"storageMode":"memory_optimized","indexerThreads":0,"memorySnapshotInterval":200,"stableSnapshotInterval":5000,"maxRollbackPoints":5,"logLevel":"info"}
```

running this this command:

```
curl -v -s -X POST $CBWOP:8091/pools/default -d memoryQuota=512
```

before running the `setupServices` curl command does the trick. Found out the other folks on chouchbase issue tracking were having the same problem: [https://issues.couchbase.com/browse/MB-16233](https://issues.couchbase.com/browse/MB-16233).

I'll be moving on to dockerizing Marklogic next.

**Comment by nicflores** [comment]
I've got Marklogic dockerized. The pre-configuration commands within the `installMarklogic` script, i.e. login to marklogic website and download of `Marklogic.rpm`, have been migrated to a `Marklogic-Dockerfile`. The post configuration now lives in a script under a docker directory of the root quasar repo along wth dditional post configuration scripts for mongo, couchbase, and postgresql. In addition I created a `setupdbs.sh` script that wraps all the post configuration scripts into a single place.

Using what I have above I ran the following tests:

```
> it/test-only **.ReadFilesSpec -- -showtimes -ex simple read  --
> it/test-only **.ReadFilesSpec -- -showtimes
```

against dockerized databases on my local machine. All tests passed but I did have to populate my own `testing.conf` file, ideally I'd like have `setupdbs.sh` do that for me.

There's one more fix I'd like to, which is to minimize the number of `RUN` commands in the Marklogic Dockerfile in order to reduce the number of layers created when building the Marklogic image.

At this point half of the first goal is complete. The next step will be to look into which tests specifically test individual connectors.

**Comment by nicflores** [comment]
Reviewed my current progress with @drostron this morning. We looked at my current `.travis.yml` refactoring and how database containers are being created and configured. We found a couple of small issues I'll be fixing today and once I have a build in my branch that times out we can re-run a build of my branch on the `quasar-analytics` repo.

**Comment by nicflores** [comment]
[PR](https://github.com/quasar-analytics/quasar/pull/2037) submitted for local scripts and travis containerization refactor. I'll be moving on to writing some sbt tasks for integration testing.

**Comment by nicflores** [comment]
At this time PR #2037 contains scripts which may help developers manage their integration testing environment.

Here is a rundown of what is possible:

1. From within the docker directory.
  a. One can run, for example, 
```
docker-compose up -d marklogic_xml mongodb_read_only couchbase
``` 
which will spin up the passed in containers. Similarly 
```
docker-compose stop marklogic_xml mongodb_read_only couchbase
```
will stop the containers. Finally 
```
docker-compose rm marklogic_xml mongodb_read_only couchbase
```
will delete these containers.

The current list of containers available to spin up are:

```
mongodb_2_6
mongodb_3_0
mongodb_read_only
mongodb_3_2
mongodb_3_4

2. Once containers are up and running some may need some extra configured. The `setupConnectors` is able to configure all running containers or specific containers. In addition one can skip the use of `docker-compose` mentioned above and use the `-u` flag along with a list of quoted space separated connectors to bring up connectors and configure them all in one command. Below is the usage output for `setupConnectors`.
```
 [14] → ./scripts/setupConnectors -h
Usage: ./scripts/setupConnectors [-h] [-a] [-c CONNECTOR-NAME]
Create and configure dockerized mongo, couchbase, marklogic, and postgresql
connectors for integration tests with Quasar.

  -h                   help (also trigged with no parameters): display this help and exit
  -a                   configure all currently running dockerized connectors
  -c CONNECTOR-NAME    configure running dockerized connector named CONNECTOR-NAME
  -u "con1 con2..."    use docker-compose up to create and configure a quoted list of connectors
```

3. After containers are up and configured `assembleTestingConf` can be used to build an 'it/testing.conf` file. Below is the usage output of this for `assembleTestingConf`.
```
 [15] → ./scripts/assembleTestingConf -h
Usage: ./scripts/assembleTestingConf [-h] [-a] [-c CONNECTOR-NAME]
Assembles Quasar integration configuration file, it/testing.conf, from currently running
containerized connectors or for the specific CONNECTOR-NAME passed in. Works for local
development and within travis-ci.

  -h                   help (also trigged with no parameters): display this help and exit
  -a                   cleans existing testing.conf and adds connector URIs to testing.conf for all currently running dockerized connectors
  -i CONNECTOR-NAME    add connector URI to existing testing.conf file
  -c CONNECTOR-NAME    cleans exisitng testing.conf and adds a connector URI for CONNECTOR-NAME currently running
  -t                   create testing.conf with all currently running dockerized connectors plus spark local connector
```

**Comment by jdegoes** [comment]
Looking good! This should all make it into documentation of the README, probably, in a new section on build/test.

**Comment by nicflores** [comment]
Can I close this issue? 

This PR #2037  contained much of the containerization work including Dockerfiles, a `docker-compose.yml` file, scripts to create and configure containers as well as a script to generate `testing.conf` file based running containers. In addition it included an updated `.travis.yml` file which leverages docker-compose and the scripts mentioned above.

The PR, #2053, contained updated Quasar README with information on how to use the new scripts.

@alissapajer and @drostron have both used the scripts and I believe they working as they should.

**Comment by zeeiyerWork** [comment]
@nicflores  - Let's discuss scope some more before closing this out.

**Comment by nicflores** [comment]
sounds good, @zeeiyerWork.

**Comment by zeeiyerWork** [comment]
After discussion with @nicflores and based on the above comments - closing this one out. since containerization is done vis-a-vis developers. There are questions around cases, where:

1. Dockerization can play a part in making installation experience smoother (i.e. installer asks if user wants a Mongo/ML/CB etc. instance and sets those up). Chat with others next week wrt this.
2. we could work on integration tests, to separate data loading out of the tests - i.e.
- A set of tests share the same instance and loaded data.
- Performance tests rely on a prior-established Mongo instance, teed up with large data sets.

But those will be recoded as separate issues.
Thanks
